In [1]:
import tensorflow as tf
from tensorflow.keras import datasets, layers, models

Loading the MNIST dataset, a collection of 60k images of handwritten digits.

In [2]:
mnist = datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

Building the model consisting of
- flattening layer
- 128 node fully connected layer with relu activation
- 20% dropout layer, and 
- 10 node output layer with softmax activation

In [3]:
model = models.Sequential([
    layers.Flatten(input_shape=(28, 28)),
    layers.Dense(128, activation='relu'),
    layers.Dropout(0.2),
    layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

Training the model.

In [4]:
model.fit(x_train, y_train, epochs=5)

model.evaluate(x_test, y_test)

Epoch 1/5
60000/60000 [==============================] - 3s 55us/sample - loss: 0.2926 - accuracy: 0.9160
Epoch 2/5
60000/60000 [==============================] - 3s 50us/sample - loss: 0.1414 - accuracy: 0.9581
Epoch 3/5
60000/60000 [==============================] - 3s 51us/sample - loss: 0.1054 - accuracy: 0.9678
Epoch 4/5
60000/60000 [==============================] - 3s 50us/sample - loss: 0.0868 - accuracy: 0.9729
Epoch 5/5
10000/10000 [==============================] - 0s 31us/sample - loss: 0.0736 - accuracy: 0.9774


[0.07358825562114361, 0.9774]

Our model acheived 97.74% accuracy on the test set! Lets see if we can improve upon this using a convolutional neural network.

Building a convolutional neural network for the same task, consisting of
- 3 convolutional layers with kernel size 3x3
- 3 max pooling layers
- a fully connected layer

In [9]:
cnn = models.Sequential([
    layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    layers.MaxPooling2D((2,2)),
    layers.Conv2D(64, (3, 3), activation='relu'),
    layers.MaxPooling2D((2,2)),
    layers.Conv2D(64, (3, 3), activation='relu'),
    layers.MaxPooling2D((2,2)),
    layers.Flatten(),
    layers.Dense(64, activation = 'relu'),
    layers.Dense(10, activation='softmax')
])

cnn.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [10]:
x_train = x_train.reshape((60000, 28, 28, 1))
x_test = x_test.reshape((10000, 28, 28, 1))

cnn.fit(x_train, y_train, epochs = 5)

cnn.evaluate(x_test, y_test)

Epoch 1/5
60000/60000 [==============================] - 19s 309us/sample - loss: 0.2110 - accuracy: 0.9347
Epoch 2/5
60000/60000 [==============================] - 19s 321us/sample - loss: 0.0751 - accuracy: 0.9775
Epoch 3/5
60000/60000 [==============================] - 20s 330us/sample - loss: 0.0542 - accuracy: 0.9836
Epoch 4/5
60000/60000 [==============================] - 21s 356us/sample - loss: 0.0420 - accuracy: 0.9868
Epoch 5/5
10000/10000 [==============================] - 1s 104us/sample - loss: 0.0459 - accuracy: 0.9875


[0.0459121120483469, 0.9875]

98.75% accuracy! nice